packages/datetime/src/dateInput.tsx

```diff
@@ -419,70 +424,55 @@ export class DateInput extends AbstractPureComponent2<DateInputProps, IDateInput
         if (e.which === Keys.ENTER) {
             const nextDate = this.parseDate(this.state.valueString);
             this.handleDateChange(nextDate, true, true);
-        } else if (e.which === Keys.TAB) {
-            this.setState({ isOpen: false });
+        } else if (e.which === Keys.TAB && e.shiftKey) {
+            // close popover on SHIFT+TAB key press
+            this.handleClosePopover();
+        } else if (e.which === Keys.TAB && this.state.isOpen) {
+            this.getKeyboardFocusableElements().shift()?.focus();
+            // necessary to prevent focusing the second focusable element
+            e.preventDefault();
         } else if (e.which === Keys.ESCAPE) {
             this.setState({ isOpen: false });
             this.inputElement?.blur();
         }
         this.safeInvokeInputProp("onKeyDown", e);
     };
 
-    private getLastTabbableElement = () => {
-        // Popover contents are well structured, but the selector will need
-        // to be updated if more focusable components are added in the future
-        const tabbableElements = this.popoverContentElement?.querySelectorAll("input, [tabindex]:not([tabindex='-1'])");
-        const numOfElements = tabbableElements?.length ?? 0;
-        // Keep track of the last focusable element in popover and add
-        // a blur handler, so that when:
-        // * user tabs to the next element, popover closes
-        // * focus moves to element within popover, popover stays open
-        const lastTabbableElement = numOfElements > 0 ? tabbableElements[numOfElements - 1] : null;
-
-        return lastTabbableElement as HTMLElement | null;
+    private getKeyboardFocusableElements = (): HTMLElement[] => {
+        const elements: HTMLElement[] = Array.from(
+            this.popoverContentElement?.querySelectorAll(
+                "button:not([disabled]),input,[tabindex]:not([tabindex='-1'])",
```

Comment:
This is a subset of the selector used in `Overlay`. I wonder if it's worth sharing that in the future